packages/react-reconciler/src/ReactFiberLane.js

```diff
@@ -365,6 +366,46 @@ export function getNextLanes(root: FiberRoot, wipLanes: Lanes): Lanes {
   return nextLanes;
 }
 
+export function markStarvedLanesAsExpired(
+  root: FiberRoot,
+  currentTime: number,
+): void {
+  // TODO: This gets called every time we yield. We can optimize by storing
+  // the earliest expiration time on the root. Then use that to quickly bail out
+  // of this function.
+
+  const pendingLanes = root.pendingLanes;
+  const eventTimes = root.eventTimes;
+
+  // Iterate through the pending lanes and check how long they've been CPU
+  // bound. After a certain amount of time, we'll assume the update is being
+  // starved and mark it as expired to force it to finish.
+  let lanes = pendingLanes;
+  while (lanes > 0) {
+    const index = ctrz(lanes);
+    const lane = 1 << index;
+
+    const eventTime = eventTimes[index];
+    if (eventTime !== -1) {
+      getHighestPriorityLanes(lanes);
+      const priority = return_highestLanePriority;
+      if (priority <= IdleHydrationLanePriority) {
+        // Anything idle priority or lower should never expire.
+      } else {
+        // User interactions should expire slightly more quickly.
+        const expirationHeuristic =
+          priority >= InputContinuousLanePriority ? 1000 : 5000;
+        if (eventTime + expirationHeuristic <= currentTime) {
+          // This lane expired
+          root.expiredLanes |= lane;
```

Comment:
What are the implications for entangled updates if we only mark a subset of lanes as expired?

Reply:
Entanglement would happen in `getNextLanes`. First we'd detect that a lane had expired, so we'd choose it to work on. Then we'd combine it with the entangled lanes before returning.

Reply:
Makes sense!